payjoin/src/receiver/mod.rs

```diff
@@ -2,28 +2,48 @@ use bitcoin::util::psbt::PartiallySignedTransaction as Psbt;
 use bitcoin::{AddressType, Script, TxOut};
 
 mod error;
+mod optional_parameters;
 
 use error::InternalRequestError;
 pub use error::RequestError;
+use optional_parameters::Params;
 
 pub trait Headers {
     fn get_header(&self, key: &str) -> Option<&str>;
 }
 
+pub trait Proposal {
+    fn is_output_substitution_disabled(&self) -> bool;
+}
+
+macro_rules! impl_proposal {
+    ($name:ident) => {
+        impl Proposal for $name {
+            fn is_output_substitution_disabled(&self) -> bool {
+                self.params.disable_output_substitution
+            }
+        }
+    };
+}
```

Comment:
Why does this need to be a trait and why it needs to be implemented for every state?

Reply:
It probably does not need to be a trait nor implemented all over. I wanted to have this available before the PSBT checks because it's the least costly check but it could probably just go on the Checked UnlockedProposal